pkg/providers/vsphere/controlplane_test.go

```diff
@@ -125,6 +125,38 @@ func TestControlPlaneSpecNoKubeVersion(t *testing.T) {
 	g.Expect(err).To(MatchError(ContainSubstring("generating vsphere control plane yaml spec")))
 }
 
+func TestControlPlaneSpecUpdateMachineTemplates(t *testing.T) {
+	g := NewWithT(t)
+	logger := test.NewNullLogger()
+	ctx := context.Background()
+	spec := givenClusterSpec(t, testClusterConfigMainFilename)
+	originalKubeadmControlPlane := kubeadmControlPlane()
+	originalEtcdCluster := etcdCluster()
+	originalEtcdCluster.Spec.InfrastructureTemplate.Name = "test-etcd-2"
+	client := test.NewFakeKubeClient(
+		originalKubeadmControlPlane,
+		originalEtcdCluster,
+		vsphereMachineTemplate("test-control-plane-1"),
+		vsphereMachineTemplate("test-etcd-2"),
+	)
+
+	kcp := wantKCP()
+	kcp.Spec.MachineTemplate.InfrastructureRef.Name = "test-control-plane-2"
+
+	etcd := wantEtcdCluster()
+	etcd.Spec.InfrastructureTemplate.Name = "test-etcd-3"
+
+	cp, err := vsphere.ControlPlaneSpec(ctx, logger, client, spec)
```

Comment:
why is the name update being triggered if we are not modifying the spec?

Reply:
I would like to see the spec being modified and those changes being reflected in the returned objects.